lib/validators/specResolver.js

```diff
@@ -657,9 +657,32 @@ class SpecResolver {
     // Adding the model name or it's discriminator value as an enum constraint with one value (constant) on property marked as discriminator
     if (definitions[name] && definitions[name].properties && definitions[name].properties[discriminator]) {
       let val = name;
+      let discriminatorType = 'string';
       if (definitions[name]['x-ms-discriminator-value']) {
         val = definitions[name]['x-ms-discriminator-value'];
       }
+      // Ensure that the property marked as a discriminator has only one enum constraint and it should be the one that we want to set as a constant.
+      if (definitions[name].properties[discriminator].enum) {
+        delete definitions[name].properties[discriminator].enum;
+      }
+      if (definitions[name].properties[discriminator]['$ref']) {
+        let obj;
+        try {
+          obj = utils.getObject(this.specInJson, definitions[name].properties[discriminator]['$ref'].slice(1));
+          if (obj && obj.type) {
+            discriminatorType = obj.type;
+          }
+        } catch (err) {
+          // do nothing, we tried to find the object and it's type. We will set it to "string" by default. It is safe to assume that properties
+          // marked as "discriminator" will be of type "string" as it needs to refer to a model definition name which logically should be a 
+          // string as it is the key in the definitions object in the swagger spec.
+        }
+        delete definitions[name].properties[discriminator]['$ref'];
+      }
+      // Set the type if not set.
+      if (!definitions[name].properties[discriminator].type) {
+        definitions[name].properties[discriminator].type = discriminatorType;
+      }
       definitions[name].properties[discriminator].enum = [`${val}`];
```

Comment:
was this line not already overwriting what the enum was, and setting it to one value? was the issue caused (for the reported case) due to having a reference for the type?

Reply:
yes the issue was caused due to the discriminator referencing an enum type defined as a separate definition.